pytential/symbolic/execution.py

```diff
@@ -711,28 +712,31 @@ def __init__(self, places, auto_where=None):
 
         # {{{ construct dict
 
-        self.places = {}
-        self.caches = {}
+        places_dict = {}
 
         auto_source, auto_target = _prepare_auto_where(auto_where)
         if isinstance(places, QBXLayerPotentialSource):
-            self.places[auto_source.geometry] = places
+            places_dict[auto_source.geometry] = places
             auto_target = auto_source
         elif isinstance(places, TargetBase):
-            self.places[auto_target.geometry] = places
+            places_dict[auto_target.geometry] = places
             auto_source = auto_target
         if isinstance(places, (Discretization, PotentialSource)):
-            self.places[auto_source.geometry] = places
-            self.places[auto_target.geometry] = places
+            places_dict[auto_source.geometry] = places
+            places_dict[auto_target.geometry] = places
         elif isinstance(places, tuple):
             source_discr, target_discr = places
-            self.places[auto_source.geometry] = source_discr
-            self.places[auto_target.geometry] = target_discr
+            places_dict[auto_source.geometry] = source_discr
+            places_dict[auto_target.geometry] = target_discr
         else:
-            self.places = places
+            places_dict = places
 
+        import immutables
+        self.places = immutables.Map(places_dict)
```

Comment:
Went ahead and made this immutable, just to be on the safe side. Should this use `pyrsistent` instead? I think `sumpy` depends on that already, right?